pkg/api/validate/dynamic/dynamic.go

```diff
@@ -197,27 +197,46 @@ func (dv *dynamic) validateVnetPermissions(ctx context.Context, vnet azure.Resou
 		"Microsoft.Network/virtualNetworks/subnets/write",
 	})
 
+	noPermissionsErr := api.NewCloudError(
+		http.StatusBadRequest,
+		errCode,
+		"",
+		"The %s service principal does not have Network Contributor permission on vnet '%s'.",
+		dv.authorizerType,
+		vnet.String(),
+	)
+
 	if err == wait.ErrWaitTimeout {
-		return api.NewCloudError(
-			http.StatusBadRequest,
-			errCode,
-			"",
-			"The %s service principal does not have Network Contributor permission on vnet '%s'.",
-			dv.authorizerType,
-			vnet.String(),
-		)
+		return noPermissionsErr
 	}
-	if detailedErr, ok := err.(autorest.DetailedError); ok &&
-		detailedErr.StatusCode == http.StatusNotFound {
-		return api.NewCloudError(
-			http.StatusBadRequest,
-			api.CloudErrorCodeInvalidLinkedVNet,
-			"",
-			"The vnet '%s' could not be found.",
-			vnet.String(),
-		)
+	if detailedErr, ok := err.(autorest.DetailedError); ok {
+		switch detailedErr.StatusCode {
+		case http.StatusNotFound:
+			return api.NewCloudError(
+				http.StatusBadRequest,
+				api.CloudErrorCodeInvalidLinkedVNet,
+				"",
+				"The vnet '%s' could not be found.",
+				vnet.String(),
+			)
+		case http.StatusForbidden:
+			noPermissionsErr.Message = fmt.Sprintf(
+				"%s Original error message: %s",
+				noPermissionsErr.Message,
+				detailedErr.Message,
+			)
+			return noPermissionsErr
+		}
 	}
-	return err
+	if err != nil { // avoiding blank Internal Server Error
+		return &api.CloudError{
```

Comment:
you don't need to return CloudError, which is only used when it's user error. returning `err` as is will result in internal server error like it has been doing.

Reply:
+1 we do not need to duplicate the logic we have to wrap uncaught errors as internal server errors. The logic can be viewed here: https://github.com/Azure/ARO-RP/blob/master/pkg/frontend/middleware/panic.go#L23

Reply:
The linked code above and [this](https://github.com/Azure/ARO-RP/blob/12aa588dbc7c95c07817bda86bb0b57774259eb7/pkg/backend/openshiftcluster.go#L256) are what I'm trying to address (see the original issue from a Hashicorp engineer [here](https://github.com/Azure/azure-rest-api-specs/issues/23355#issuecomment-1522415332).  When it is not a cloud error, the original message is lost and only "Internal Server Error" is sent back to the caller.

What I had in mind was most unexpected/uncaught errors during the validation were interactions with ARM, so it might be ok to package the original error message and return to the caller so that they have an idea about what's happening.  Alternatively, we could make the locations above include the original message, but I'm not so sure if that's a good idea.. if we're not careful, it could send irrelevant or even leak some sensitive info.

What do you think about this?  I'm open to suggestions here.

Reply:
Unwrapped now.